Automattic/wp-calypso
# Themes: Use `details` middleware for server-side rendering

In the `/themes` route in `server/pages`, use `my-sites/themes` to create a theme sheet element tree in `context.layout`, and `renderToString()` that.
 * Adapt `express`'s `req` context to be able to use it with middleware that otherwise operates on `page`'s `context`. props @mcsf
 * Add a `package.json` that sets different entry points on the server and in the browser (not everything required by `my-sites/themes/index.js` is SSR-ready yet, so we need a pruned version for the server).

* Move theme details specific async data fetching to a separate middleware in  `my-sites/themes/controller/index.node.js`.
* Move generic `renderToString` caching to `server/render`.
* Cache Helmet data, so `title` etc are properly set.

To test:
* Check that everything still works as before.
* Try logged-out `/theme/mood`, with JS disabled. Note that the title is now set to '`themeName` Theme — WordPress.com' (instead of the generic 'WordPress.com', as on `master`).
* Note that logged-in `/design` and `/themes/something` routes server-render the logged-out layout (masterbar!) before it's being replaced by the logged-in one on the client. While we want to fix this in the long run, it is no behavioral change from `master`.

Note:
* There are still reconciliation errors, but there are also some in `master`. Opting to fix them later, since this PR is big enough already. (Possible cuplrit: Apparently, `react-helmet` inserts a `noscript` tag. https://github.com/nfl/react-helmet/issues/81)
* SEO is still broken for SSR'd `/design`. Also opting to fix later.

/cc @seear @ehg @mcsf

Follow-ups: #3302, #3774, #3795, #3425 

I'd say this closes #185 :-)
